### ISSUE TITLE
com.hazelcast.jet.impl.connector.WriteFilePTest.stressTest_exactlyOnce_forceful 

### ISSUE THREAD

**Issue by viliam-durina** [hazelcast/hazelcast-jet#2504]
Moved from #2487

Failed on Zulu JDK 11 but with different exception: http://jenkins.hazelcast.com/job/jet-oss-master-zulu-jdk11/261/testReport/junit/com.hazelcast.jet.impl.connector/WriteFilePTest/stressTest_exactlyOnce_forceful/

Seems like this issue which occurred on the maintenance branch some time ago: https://github.com/hazelcast/hazelcast-jet/issues/2277

Stacktrace:
```
org.junit.ComparisonFailure: 
expected:<
0
1
2
3
...> but was:<
0
0
1
1
2
2
3
3
...>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.hazelcast.jet.impl.connector.WriteFilePTest.checkFileContents(WriteFilePTest.java:437)
	at com.hazelcast.jet.impl.connector.WriteFilePTest.stressTest(WriteFilePTest.java:413)
	at com.hazelcast.jet.impl.connector.WriteFilePTest.stressTest_exactlyOnce_forceful(WriteFilePTest.java:365)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Standard output:
```
Hiccups measured while running test 'stressTest_exactlyOnce_forceful(com.hazelcast.jet.impl.connector.WriteFilePTest):'
03:21:35, accumulated pauses: 486 ms, max pause: 162 ms, pauses over 1000 ms: 0
03:21:40, accumulated pauses: 1228 ms, max pause: 404 ms, pauses over 1000 ms: 0
03:21:45, accumulated pauses: 252 ms, max pause: 20 ms, pauses over 1000 ms: 0
03:21:50, accumulated pauses: 420 ms, max pause: 145 ms, pauses over 1000 ms: 0
03:21:55, accumulated pauses: 365 ms, max pause: 101 ms, pauses over 1000 ms: 0
03:22:00, accumulated pauses: 414 ms, max pause: 86 ms, pauses over 1000 ms: 0
03:22:05, accumulated pauses: 65 ms, max pause: 6 ms, pauses over 1000 ms: 0
03:22:10, accumulated pauses: 130 ms, max pause: 31 ms, pauses over 1000 ms: 0
03:22:15, accumulated pauses: 57 ms, max pause: 10 ms, pauses over 1000 ms: 0
03:22:20, accumulated pauses: 83 ms, max pause: 54 ms, pauses over 1000 ms: 0
03:22:25, accumulated pauses: 55 ms, max pause: 5 ms, pauses over 1000 ms: 0
03:22:30, accumulated pauses: 46 ms, max pause: 4 ms, pauses over 1000 ms: 0
03:22:35, accumulated pauses: 97 ms, max pause: 62 ms, pauses over 1000 ms: 0
```
The original test output: [com.hazelcast.jet.impl.connector.WriteFilePTest-output.txt.gz](https://github.com/hazelcast/hazelcast-jet/files/5183881/com.hazelcast.jet.impl.connector.WriteFilePTest-output.txt.gz)

**Comment by viliam-durina** [comment]
TLDR: First 350 items out of 500 were duplicated. All logs seem reasonable, can't reproduce it locally, didn't find anything strange in the logs.

Investigation: Surefire tries to be smart with the `assertEquals` failure: it truncates the expected and actual parts and displays the different part in `[]`. So we know that items 0..350 are duplicated, items 351..499 are not.

Reviewing the logs I couldn't find a cause. It doesn't reproduce locally, nor did it fail before. The job restarted many times, but I didn't find that the job could be running twice. We don't have the original files created by the test. The test writes to a directory created using `Files.createTempDirectory()`, it's separate for each test. For this test the actual directory was `/tmp/write-file-p1834102381601195356`. I checked all other tests and none logged the value `p1834102381601195356`.

The assert has a 60-second timeout, this is the time after which the test failed. It writes 500 items, sleeping 5ms after each item, which makes at most 200 items/s. In this execution the emission took 9 seconds, as seen from the sink logs when it stopped creating more items, which seems reasonable: 200 items/s should take 5 secs, but takes longer due to the forceful restarts and reemissions of items after restore.

I also checked that the job accidentally didn't join another cluster, but all messages are logged from the same member. Also, the source is non-distributed. The edge from source to sink is round-robin, local.

This is the logged DAG:
```
digraph DAG {
	"src" [localParallelism=1];
	"filesSink(/tmp/write-file-p1834102381601195356)" [localParallelism=2];
	"src" -> "filesSink(/tmp/write-file-p1834102381601195356)" [queueSize=1024];
}
```

**Comment by viliam-durina** [comment]
Further idea is that after emitting 350 items, the source loses the state and starts again.

Suspected logs were found after 2020-09-06 03:21:40,582:
- snapshot 17 was started
- job was forcefully restarted, the snapshot 17 failed in phase 1 before storing anything
- the execution was terminated
- the response to phase 1 for snapshot 17 was handled, map `__jet.snapshot.04ec-9a7b-cec4-0001.0` was cleared
- then the job was started again and restored from snapshot 16 stored in `__jet.snapshot.04ec-9a7b-cec4-0001.1`

So this looked similar to the issue fixed in #2503, but it's not.

I'll continue checking the log tomorrow.

**Comment by olukas** [comment]
Failed again on master with Zulu JDK 8: http://jenkins.hazelcast.com/job/jet-oss-master-zulu-jdk8/322/testReport/junit/com.hazelcast.jet.impl.connector/WriteFilePTest/stressTest_exactlyOnce_forceful/
